pkg/operator/defragcontroller/defragcontroller.go

```diff
@@ -0,0 +1,188 @@
+			Type:   "DefragControllerDegraded",
+			Status: operatorv1.ConditionFalse,
+			Reason: "AsExpected",
+		}))
+	return updateErr
+}
+
+func (c *DefragController) checkDefrag(ctx context.Context, recorder events.Recorder) error {
+	controlPlaneTopology, err := ceohelpers.GetControlPlaneTopology(c.infrastructureLister)
+	if err != nil {
+		return err
+	}
+	// defrag is blocking and can only be safely performed in HighlyAvailableTopologyMode
+	if controlPlaneTopology != configv1.HighlyAvailableTopologyMode {
+		klog.V(4).Infof("DefragController is disabled for control plane topology: %s", controlPlaneTopology)
+		return nil
+	}
+
+	etcdMembers, err := c.etcdClient.MemberList()
+	if err != nil {
+		return err
+	}
+
+	// dont defrag if any of the cluster members are unhealthy
+	memberHealth, err := c.etcdClient.MemberHealth()
+	if err != nil {
+		return err
+	}
+	if !etcdcli.IsClusterHealthy(memberHealth) {
+		return fmt.Errorf("cluster is unhealthy: %s", memberHealth.Status())
+	}
+
+	var endpointStatus []*clientv3.StatusResponse
+	var leader *clientv3.StatusResponse
+	for _, member := range etcdMembers {
+		status, err := c.etcdClient.Status(ctx, member)
+		if err != nil {
+			return err
+		}
+		if leader == nil && status.Leader == member.ID {
+			leader = status
+			continue
+		}
+		endpointStatus = append(endpointStatus, status)
+	}
+
+	// leader last if possible
+	if leader != nil {
+		endpointStatus = append(endpointStatus, leader)
+	}
+
+	var errors []error
+	for _, status := range endpointStatus {
+		member, err := getMemberFromStatus(etcdMembers, status)
+		if err != nil {
+			errors = append(errors, err)
+		}
+
+		// Check each member's status which includes the db size on disk "DbSize" and the db size in use "DbSizeInUse"
+		// compare the % difference and if that difference is over the max diff threshold and also above the minimum
+		// db size we defrag the members state file. In the case where this command only partially completed controller
+		// can clean that up on the next sync. Having the db sizes slightly different is not a problem in itself.
+		if isEndpointFragmented(member, status) {
+			recorder.Eventf("DefragControllerDefragmentAttempt", "Attempting defrag on member: %s, dbSize: %d , dbInUse: %d, memberID %d, leader ID: %d", member.Name, status.DbSize, status.DbSizeInUse, member.ID, status.Leader)
+			if _, err := c.etcdClient.Defragment(ctx, member); err != nil {
+				errors = append(errors, err)
+			}
+			postDefragEndpointStatus, err := c.etcdClient.Status(ctx, member)
+			if err != nil {
+				errors = append(errors, err)
+			}
+
+			recorder.Eventf("DefragmentSuccess", "etcd member has been defragmented: %s, dbSize: %d, dbInUse: %d, member ID: %d, leader ID: %d", member.Name, postDefragEndpointStatus.DbSize, postDefragEndpointStatus.DbSizeInUse, member.ID, postDefragEndpointStatus.Leader)
+
+			//give cluster time to recover before we move to the next member.
+			if err := wait.Poll(
+				waitDuration,
```

Comment:
Is two seconds before we query for member health enough for us to determine this?

Reply:
I think so that is 2x election timeout, plus we have a retry until 30s if nessisary.